src/lustre/lustreGenRefTypeImpNodes.ml

```diff
@@ -0,0 +1,148 @@
+                                      ty_str in
+    i := !i + 1;
+    let is_extern = true in
+    let params = [] in 
+    let vars = AH.vars_without_node_call_ids expr in
+    let inputs = SI.diff vars (SI.singleton id) |> SI.elements in
+    let inputs = List.filter_map (fun id -> 
+      let ty = Ctx.lookup_ty ctx id |> unwrap in
+      let ty = Chk.expand_type_syn_reftype_history ctx ty |> unwrap_res in
+      let is_const = Ctx.member_val ctx id in
+      let call_params = (match node_id with 
+      | Some node_id -> Ctx.lookup_node_param_ids ctx node_id |> unwrap 
+      | None -> []
+      ) in
+      let is_global_const = is_const && not (List.mem id call_params) in
+      if is_global_const 
+      then None 
+      else Some (pos, id, ty, A.ClockTrue, is_const)
+    ) inputs in
+    let outputs = [(pos, id, ty, A.ClockTrue)] in
+    let locals = [] in 
+    let node_items = [A.AnnotMain(pos, true)] in 
+    (* Add assumption for each variable with a refinement type in 'expr' *)
+    let assumptions = List.filter_map (fun (_, id, _, _, _) -> 
+      let ty = Ctx.lookup_ty ctx id |> unwrap in 
+      match ty with 
+        | A.RefinementType (_, (_, id2, _), expr) -> 
+          let expr =  (AH.substitute_naive id2 (Ident(pos, id)) expr) in
+          Some (A.Assume (pos, None, false, expr))
+        | _ -> None 
+    ) inputs in 
+    (* Add guarantee for 'expr' *) 
+    let guarantee = A.Guarantee (pos, None, false, expr) in
+    let contract = Some (guarantee :: assumptions) in
+    Some (NodeDecl (span, (gen_node_id, is_extern, params, inputs, outputs, locals, node_items, contract)))
+  | _ -> None
+
+let gen_imp_nodes: Ctx.tc_context -> A.declaration list -> A.declaration list 
+= fun ctx decls -> 
+  List.fold_left (fun acc decl -> 
+    match decl with 
+    | A.TypeDecl (_, AliasType (_, _, ty))
+    | A.ConstDecl (_, FreeConst (_, _, ty))
+    | A.ConstDecl (_, TypedConst (_, _, _, ty)) -> 
+      (match (ref_type_to_contract ctx ty None) with  
+      | None -> decl :: acc
+      | Some decl2 -> decl :: decl2 :: acc)
+    | A.TypeDecl (_, FreeType _)
+    | A.ConstDecl (_, UntypedConst _) -> decl :: acc
+    | A.NodeDecl (span, ((id, extern, params, inputs, outputs, locals, node_items, contract) as decl2)) -> 
+      let decl = 
+        if not (List.mem `CONTRACTCK (Flags.enabled ())) then decl else
+        (* To prevent slicing, we mark generated imported nodes as main nodes *)
+        A.NodeDecl(span, (id, extern, params, inputs, outputs, locals, AnnotMain(Lib.dummy_pos, true) :: node_items, contract)) 
+      in
+      (match node_decl_to_contracts decl2 with 
+      | None -> decl :: acc
+      | Some (decl2, decl3) -> decl :: decl2 :: decl3 :: acc)
+    | A.FuncDecl (span, ((id, extern, params, inputs, outputs, locals, node_items, contract) as decl2)) -> 
+      let decl = 
+        if not (List.mem `CONTRACTCK (Flags.enabled ())) then decl else
```

Comment:
My understanding is that the generation of imported nodes associated with refinement types is only necessary when `List.mem CONTRACTCK (Flags.enabled ())` is true, right? If that is the case, I would remove all these checks in this file, only call to `gen_imp_nodes` in `lustreInput` if the condition is true, and assume the condition holds in the function has been called.

Reply:
Addressed